File: sdk/react-native/ios/TraceMindModule.swift
#if canImport(React)
import Foundation
import React
import TraceMind

@objc(TraceMindModule)
final class TraceMindModule: NSObject {
  @objc static func requiresMainQueueSetup() -> Bool {
    true
  }

  @objc func start(_ config: NSDictionary) {
    guard let projectKey = config["projectKey"] as? String else { return }
    if let endpointValue = config["endpoint"] as? String, let endpoint = URL(string: endpointValue) {
      TraceMind.start(projectKey: projectKey, endpoint: endpoint)
    } else {
      TraceMind.start(projectKey: projectKey)
    }
  }

  @objc func capture(_ type: String, payload: NSDictionary) {
    let eventName = payload["eventName"] as? String
    let path = payload["path"] as? String ?? "ReactNative"
    let properties = Self.traceMindFields(payload["properties"])
    let context = Self.traceMindFields(payload["context"])
    try? TraceMind.capture(type, eventName: eventName, path: path, properties: properties, context: context)
  }

  @objc func identify(_ userId: String, traits: NSDictionary) {
    try? TraceMind.identify(userId, traits: Self.traceMindFields(traits))
  }

  @objc func setScreen(_ screen: String) {
    TraceMind.setScreen(screen)
  }

  @objc func setAttribution(_ attribution: NSDictionary) {
    TraceMind.setAttribution(Self.traceMindAttribution(attribution))
  }

  @objc func recordDeepLink(_ payload: NSDictionary) {
    guard let urlValue = payload["url"] as? String, let url = URL(string: urlValue) else { return }
    let sourceApplication = payload["sourceApplication"] as? String
      ?? payload["sourcePackage"] as? String
      ?? payload["referrer"] as? String
    TraceMind.recordOpenURL(url, sourceApplication: sourceApplication)
  }

  @objc func submitFeedback(_ payload: NSDictionary) {
    guard let messageInput = payload["message"] as? NSDictionary else { return }
    let contactInput = messageInput["contact"] as? NSDictionary
    let contact = TraceMindFeedbackContact(
      name: contactInput?["name"] as? String,
      email: contactInput?["email"] as? String,
      phone: contactInput?["phone"] as? String,
      preferredChannel: contactInput?["preferredChannel"] as? String,
      consent: contactInput?["consent"] as? Bool ?? false
    )
    let message = TraceMindFeedbackMessage(
      kind: messageInput["kind"] as? String ?? "other",
      title: messageInput["title"] as? String,
      body: messageInput["body"] as? String ?? "",
      contact: contact,
      fields: Self.traceMindFields(messageInput["fields"])
    )
    Task {
      try? await TraceMind.submitFeedback(
        message: message,
        path: payload["path"] as? String,
        title: payload["title"] as? String
      )
    }
  }

  private static func traceMindFields(_ value: Any?) -> TraceMindFields {
    guard let fields = value as? NSDictionary else { return [:] }
    var next: TraceMindFields = [:]
    fields.forEach { key, value in
      guard let key = key as? String else { return }
      if let stringValue = value as? String {
        next[key] = .string(stringValue)
        return
      }
      if let numberValue = value as? NSNumber {
        if CFGetTypeID(numberValue) == CFBooleanGetTypeID() {
          next[key] = .bool(numberValue.boolValue)
        } else {
          next[key] = .number(numberValue.doubleValue)
        }
      }
    }
    return next
  }

  private static func traceMindAttribution(_ value: NSDictionary) -> TraceMindAttribution {
    TraceMindAttribution(
      source: value["source"] as? String,
      medium: value["medium"] as? String,
      campaign: value["campaign"] as? String,
      content: value["content"] as? String,
      referrerDomain: value["referrerDomain"] as? String,
      referrerType: value["referrerType"] as? String,
      landingPath: value["landingPath"] as? String,
      gclidPresent: boolValue(value["gclidPresent"]),
      fbclidPresent: boolValue(value["fbclidPresent"]),
      msclkidPresent: boolValue(value["msclkidPresent"])
    )
  }

  private static func boolValue(_ value: Any?) -> Bool? {
    if let value = value as? Bool { return value }
    if let value = value as? NSNumber { return value.boolValue }
    return nil
  }
}
#endif

File: sdk/react-native/android/src/main/java/com/tracemind/reactnative/TraceMindModule.kt
package com.tracemind.reactnative

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.tracemind.TraceMind
import com.tracemind.TraceMindFeedbackContact
import com.tracemind.TraceMindFeedbackMessage

class TraceMindModule(
  private val reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext) {
  override fun getName(): String = "TraceMindModule"

  @ReactMethod
  fun start(config: ReadableMap) {
    val projectKey = config.getString("projectKey") ?: return
    TraceMind.start(
      application = reactContext.applicationContext as android.app.Application,
      projectKey = projectKey,
      endpoint = if (config.hasKey("endpoint")) config.getString("endpoint") ?: DEFAULT_ENDPOINT else DEFAULT_ENDPOINT
    )
  }

  @ReactMethod
  fun capture(type: String, payload: ReadableMap) {
    TraceMind.capture(
      type = type,
      eventName = if (payload.hasKey("eventName")) payload.getString("eventName") else null,
      path = if (payload.hasKey("path")) payload.getString("path") ?: "ReactNative" else "ReactNative",
      properties = readPrimitiveMap(payload, "properties"),
      context = readPrimitiveMap(payload, "context")
    )
  }

  @ReactMethod
  fun identify(userId: String, traits: ReadableMap) {
    TraceMind.identify(userId = userId, traits = readPrimitiveMap(traits))
  }

  @ReactMethod
  fun setScreen(screen: String) {
    TraceMind.setScreen(screen)
  }

  @ReactMethod
  fun submitFeedback(payload: ReadableMap) {
    val messageMap = if (payload.hasKey("message")) payload.getMap("message") else null
    val contactMap = messageMap?.let { if (it.hasKey("contact")) it.getMap("contact") else null }
    TraceMind.submitFeedback(
      message = TraceMindFeedbackMessage(
        kind = messageMap?.let { if (it.hasKey("kind")) it.getString("kind") else null } ?: "other",
        title = messageMap?.let { if (it.hasKey("title")) it.getString("title") else null },
        body = messageMap?.let { if (it.hasKey("body")) it.getString("body") else null } ?: "",
        contact = TraceMindFeedbackContact(
          name = contactMap?.let { if (it.hasKey("name")) it.getString("name") else null },
          email = contactMap?.let { if (it.hasKey("email")) it.getString("email") else null },
          phone = contactMap?.let { if (it.hasKey("phone")) it.getString("phone") else null },
          preferredChannel = contactMap?.let { if (it.hasKey("preferredChannel")) it.getString("preferredChannel") else null },
          consent = contactMap?.let { it.hasKey("consent") && it.getBoolean("consent") } ?: false
        ),
        fields = messageMap?.let { readPrimitiveMap(it, "fields") } ?: emptyMap()
      ),
      path = if (payload.hasKey("path")) payload.getString("path") ?: "ReactNative" else "ReactNative",
      title = if (payload.hasKey("title")) payload.getString("title") else null
    )
  }

  private fun readPrimitiveMap(payload: ReadableMap, key: String): Map<String, Any?> {
    if (!payload.hasKey(key)) return emptyMap()
    return payload.getMap(key)?.let { readPrimitiveMap(it) } ?: emptyMap()
  }

  private fun readPrimitiveMap(fields: ReadableMap): Map<String, Any?> {
    val next = mutableMapOf<String, Any?>()
    val iterator = fields.keySetIterator()
    while (iterator.hasNextKey()) {
      val key = iterator.nextKey()
      if (fields.isNull(key)) continue
      when (fields.getType(key)) {
        ReadableType.String -> next[key] = fields.getString(key)
        ReadableType.Number -> next[key] = fields.getDouble(key)
        ReadableType.Boolean -> next[key] = fields.getBoolean(key)
        else -> Unit
      }
    }
    return next
  }

  private companion object {
    const val DEFAULT_ENDPOINT = "https://tracemind.sandbox.galaxycloud.app/api/capture"
  }
}
